relay-general/src/protocol/contexts/device.rs

```diff
@@ -169,9 +172,80 @@ impl DeviceContext {
     }
 }
 
-#[test]
-fn test_device_context_roundtrip() {
-    let json = r#"{
+impl FromUserAgentInfo for DeviceContext {
+    fn from_client_hints(client_hints: &ClientHints) -> Option<Self> {
+        let device = client_hints.sec_ch_ua_model?.to_owned();
+        Some(Self {
+            family: Annotated::new(device),
+            ..Default::default()
+        })
+    }
+
+    fn from_user_agent(user_agent: &str) -> Option<Self> {
+        let device = parse_device(user_agent);
+
+        if !is_known(device.family.as_str()) {
```

Comment:
Will this work if you use following ?
```suggestion
        if !is_known(&device.family) {
```